chalice/watcher/stat.py

```diff
@@ -0,0 +1,66 @@
+import threading
+import time
+
+from typing import Dict, List, Set, Callable  # noqa
+
+from chalice.watcher.shared import Watcher
+from chalice.utils import OSUtils
+
+
+class StatFileObserver(object):
+    def __init__(self, path):
+        # type: (str) -> None
+        self._path = path
+        self._osutils = OSUtils()
+        self._mtimes = {}  # type: Dict[str, int]
+
+    def check(self):
+        # type: () -> Set[str]
+        return self._check_dir(self._path)
+
+    def _check_dir(self, directory):
```

Comment:
Out of curiosity, I compared the cpu usage of what flask has vs. this implementation, and trying to keep things similar in terms of files being watched I noticed substantially higher CPU usage here.  With werkzeug I see:

```
# Werkzeug
Total files watched: 277
CPU:
$ while true; do ps aux | grep [5]5057; sleep 1; done
USER    PID  %CPU %MEM      VSZ    RSS   TT  STAT STARTED      TIME COMMAND
j     55057   2.9  0.1  2483948  24788 s003  S+    3:28PM   0:01.43 ...flask run
j     55057   2.2  0.1  2483948  24788 s003  S+    3:28PM   0:01.45 ...flask run
j     55057   1.4  0.1  2483948  24788 s003  S+    3:28PM   0:01.46 ...flask run
j     55057   0.9  0.1  2483948  24788 s003  S+    3:28PM   0:01.47 ...flask run
j     55057   0.9  0.1  2483948  24788 s003  S+    3:28PM   0:01.49 ...flask run
j     55057   0.5  0.1  2483948  24788 s003  S+    3:28PM   0:01.50 ...flask run
j     55057   0.5  0.1  2483948  24788 s003  S+    3:28PM   0:01.51 ...flask run
j     55057   0.3  0.1  2483948  24788 s003  S+    3:28PM   0:01.52 ...flask run
j     55057   0.3  0.1  2483948  24788 s003  S+    3:28PM   0:01.54 ...flask run
j     55057   0.2  0.1  2483948  24788 s003  S+    3:28PM   0:01.55 ...flask run
j     55057   0.1  0.1  2483948  24788 s003  S+    3:28PM   0:01.56 ...flask run
j     55057   0.1  0.1  2483948  24788 s003  R+    3:28PM   0:01.58 ...flask run
j     55057   3.7  0.1  2483948  24788 s003  S+    3:28PM   0:01.60 ...flask run
j     55057   3.8  0.1  2483948  24788 s003  S+    3:28PM   0:01.61 ...flask run
j     55057   2.5  0.1  2483948  24788 s003  S+    3:28PM   0:01.63 ...flask run
j     55057   1.7  0.1  2483948  24788 s003  S+    3:28PM   0:01.64 ...flask run
j     55057   1.1  0.1  2483948  24788 s003  S+    3:28PM   0:01.65 ...flask run
```

werkzeug checks files from `sys.modules` vs. we check all files recursively, so I tried to create a similarly sized app dir:

```
# Chalice stat() version
$ find . | wc -l
     315

$ while true; do ps aux | grep [5]6087; sleep 1; done
USER    PID  %CPU %MEM      VSZ    RSS   TT  STAT STARTED      TIME COMMAND
j     56087  94.1  0.2  2495124  33280 s001  R+    3:37PM   0:11.31 chalice local
j     56087  11.2  0.2  2495124  33280 s001  R+    3:37PM   0:11.42 chalice local
j     56087  32.1  0.2  2495124  33280 s001  S+    3:37PM   0:12.21 chalice local
j     56087  71.1  0.2  2495124  33280 s001  R+    3:37PM   0:12.62 chalice local
j     56087  10.2  0.2  2495124  33280 s001  S+    3:37PM   0:13.08 chalice local
j     56087  93.5  0.2  2495124  33280 s001  R+    3:37PM   0:13.83 chalice local
j     56087   5.1  0.2  2495124  33280 s001  R+    3:37PM   0:13.96 chalice local
j     56087  97.6  0.2  2495124  33280 s001  S+    3:37PM   0:14.98 chalice local
j     56087  31.5  0.2  2495124  33280 s001  R+    3:37PM   0:15.17 chalice local
j     56087  25.8  0.2  2495124  33280 s001  S+    3:37PM   0:15.86 chalice local
j     56087  86.7  0.2  2495124  33280 s001  R+    3:37PM   0:16.39 chalice local

```

While this doesn't have to be the most performance code, I don't think we should be eating up 90%+ cpu in pretty reasonable cases, especially when other frameworks don't seem to have this issue monitoring similar file sizes.

There some small issues I see (such as we creating an entire list of changed files when really we only care if a single file has changed, we don't need to build up the full list), but that wouldn't explain the idle case here where we're taking up a lot of CPU.